docs: PDR for junit4 and vavr (#61)
CSC-PS-S2/s2-workflow: docs/pdr/PDR003-vavr.md

type: tool
tags: vavr,collection
date: 2021-12-01
status: wip

---

# Problem

The code we write is destined to be used in both reactive and non-reactive
environments. Spring Cloud Function (on which Spring Cloud Stream and Spring
Cloud Dataflow rest) provides several models of development but the reactive
model allows more flexibility than the mere "functional" model.

In a reactive environment, the threading model is mostly managed by Reactor:

> the developer writes Fluxes and Monos, but it is generally up to the final

Reviewer: Maybe 
```suggestion
the developer writes `Flux` and `Mono`, but it is generally up to the final
```

to not be modified, however it is very impractical. In order to create
a new collection from an unmodifiable one, the developer has to create
a whole new collection, not reusing any sub-structure from the initial
collections. Also, when given a value for the e.g. `List` interface, one has
no way to know if it is modifiable or not from its type only.

+ Pros:
  - java collections have a well-known API,
+ Cons:
  - java collections have a rather lacking API,
  - no dedicated immutable collection API.

## Using Vavr collections

Vavr provides several implementations of purely functional collections.
These are immutable collections with different types of properties. If the
size of the collection is very small, pretty much any implementation will
do in any case. If the size of the collection can become big, it may become
crucial to decide which type of collection to use. The default implementation
for list-like collections should be `Vector`, which has all operations in
effectively constant time. The default implementation for set-like collections
should be Vavr's `HashSet`. However, in some cases other implementations

> are more benefiial.

Reviewer: ```suggestion
are more beneficial.
```

### List

These are linked-list, to be used when only iterating over the elements

> of the collection, and possibly only preprending to it. (Appending to List

Reviewer: ```suggestion
of the collection, and possibly only prepending to it. (Appending to List
```

is linear, does not modify the initial list but creates a new one.)

### Queue

These are couples of linked-lists, to be used when iterating over the elements

> of the collection, and possibly appending or preprending to it.

Reviewer: ```suggestion
of the collection, and possibly appending or prepending to it.
```

### Try/Option

> Also, vavr provides `Option` whih is a better alternative for null-safety

Reviewer: ```suggestion
Also, vavr provides `Option` which is a better alternative for null-safety
```
